### PR TITLE
[r] Enhance support for (partial) table extraction

### PR DESCRIPTION
**Pull request by eddelbuettel**
This PR adds support for access via the `SOMAReader` class to extract selected columns in a single pass. Query condition and range support will be added next.

The retrieved data structure is a from an Arrow-support package that works only on the lighterweight C interface to Arrow without linking.  Once data has been accessed, we can easily transfer to different Arrow data structures:

```r
> library(tiledbsoma)
> uri <- "test/soco/pbmc3k_processed/obs"  # local data set
> columns <- c("n_counts", "n_genes", "louvain")
> z <- export_recordbatch(uri, columns)
[2022-10-07 18:00:12.785] [tiledbsoma] [Process: 66049] [Thread: 66049] [info] Reading from test/soco/pbmc3k_processed/obs
[2022-10-07 18:00:12.795] [tiledbsoma] [Process: 66049] [Thread: 66049] [info] Read complete with 2638 obs and 3 cols
[2022-10-07 18:00:12.795] [tiledbsoma] [Process: 66049] [Thread: 66049] [info] Accessing n_counts at 0
[2022-10-07 18:00:12.795] [tiledbsoma] [Process: 66049] [Thread: 66049] [info] Accessing n_genes at 1
[2022-10-07 18:00:12.795] [tiledbsoma] [Process: 66049] [Thread: 66049] [info] Accessing louvain at 2
> 
> rb <- arch::from_arch_array(z, arrow::RecordBatch)   ## Arrow RecordBatch
> rb
RecordBatch
2638 rows x 3 columns
$ <float not null>
$ <int64 not null>
$ <large_binary not null>
> 
> tb <- arrow::as_arrow_table(arch::from_arch_array(z, arrow::RecordBatch))  ## Arrow Table
> tb
Table
2638 rows x 3 columns
$ <float not null>
$ <int64 not null>
$ <large_binary not null>
> 
```

### CODE REVIEW

**Comment by eddelbuettel** [comment]
(rebased and force-pushed)

File: apis/r/src/rinterface.cpp [hunk 1]
```diff
@@ -43,3 +59,136 @@ bool export_column(const std::string& uri, const std::string& name, SEXP schemax
...
+// [[Rcpp::export]]
+SEXP export_recordbatch(const std::string& uri, const std::vector<std::string>& colnames) {
+
+    tdbs::LOG_INFO(fmt::format("Reading from {}", uri));
+
+    // Read selected columns from the obs array (obs is unique_ptr<SOMAReader>)
+    auto obs = tdbs::SOMAReader::open(uri, "", {}, colnames);
+    obs->submit();
+
+    // Getting next batch:  std::optinal<std::shared_ptr<ArrayBuffers>>
```
Comment (johnkerl) [inline apis/r/src/rinterface.cpp:144]: nit: optinal -> optional

File: apis/r/src/rinterface.cpp [hunk 2]
```diff
@@ -43,3 +59,136 @@ bool export_column(const std::string& uri, const std::string& name, SEXP schemax
...
+                                      Rcpp::wrap(2),      	// flags, 2 == unordered, nullable, no sorted map keys
+                                      schlst, 	        	// children
+                                      R_NilValue);        	// dictionary
+    SEXP axp = arch_c_array_from_sexp(Rcpp::List::create(Rcpp::Named("")=R_NilValue), // buffers
+                                      Rcpp::wrap(rows), 	// length
+                                      Rcpp::wrap(-1), 	    // null count, -1 means not determined
+                                      Rcpp::wrap(0),    	// offset (in bytes)
+                                      arrlst,               // children
+                                      R_NilValue);          // dictionary
+    SEXP as = array_sexp_new(sxp, axp);
```
Comment (johnkerl) [inline apis/r/src/rinterface.cpp:192]: "comment gold" here 😁 ✨

File: apis/r/src/rinterface.cpp (hunk 1 above)
Comment (johnkerl) [inline apis/r/src/rinterface.cpp:144]: nit: optinal -> optional

**Review by johnkerl — APPROVED** [review]
LGTM

File: apis/r/src/rinterface.cpp (hunk 1 above)
Comment (eddelbuettel) [reply apis/r/src/rinterface.cpp]: Thanks, now fixed.

**Pull request merged**